WordPress/gutenberg
# Deactivate plugin versions below 4.1 on WordPress 5.0

~**DO NOT MERGE.** This is meant to be a patch applied directly to previous releases. Read on.~
**Edit:** Patches released tagged. See https://github.com/WordPress/gutenberg/pull/11375#issuecomment-435220976

Ignore the diff against `master`. [See actual diff against v4.0.0.](https://github.com/WordPress/gutenberg/compare/v4.0.0...fix/deactivate-if-5-0)

## Description

Core ticket [45123](https://core.trac.wordpress.org/ticket/45123) took care of systematically deactivating Gutenberg upon upgrading to WordPress 5.0, both as a general precaution and due to the fact that core WordPress becomes the default provider of Gutenberg functionality.

By "general precaution" it is understood that, although core and plugin are expected to be compatible starting with Gutenberg 4.1, they naturally share a large surface and we want to avoid any chance of failure during a core upgrade process.

However, versions of the plugin prior to 4.1 are fundamentally incompatible with WordPress 5.0, to the point where [core's procedure of deactivating](https://core.trac.wordpress.org/browser/branches/5.0/src/wp-admin/includes/upgrade.php?rev=43765#L1802) the plugin happens too late to guarantee isolation, [potentially causing a site to WSOD](https://core.trac.wordpress.org/ticket/45157).

Thus, this PR proposes a patch for Gutenberg versions 3.6, 3.7, 3.8, 3.9 and 4.0 that deactivates itself if any version of WordPress above 4.9.8 is detected. It can do this before execution of any other code in the plugin, thus ensuring no Gutenberg code taints core's upgrade routine. With the help of @pento, patch releases will be pushed as auto-updates to existing installations.

The commit list shows an initial version that showed a deactivation notice, but in practice I found this to be useless.

## How has this been tested?
- Build the plugin from this patch (`bin/build-plugin-zip.sh`).
- Install and activate it on a WP 4.9.8 site.
- Confirm that the plugin stays activated and that the Gutenberg editor is present.
- Using the [Beta Tester Plugin](https://wordpress.org/plugins/wordpress-beta-tester/) or any comparable means, upgrade to a WP 5.0 nightly or beta.
- Ensure that the process goes smoothly and that it deactivates the Gutenberg plugin.

- Repeat the steps using a *non-patched* version of 4.0 or lower and confirm that the upgrade process is negatively affected by Gutenberg.
